### PR TITLE
Updating validator and adding a tool executor

### PR DESCRIPTION
**Pull request by ttaa9**
To start addressing #19 and #17.

- The validator now checks for some consistency with the schema (including the updates in boutiques/schema#40)
- The local tool executor can take inputs in csv or json (or as a command-line string), check them for correctness against the descriptor, and then either print or execute the generated command line. It can also generate a random set of consistent input parameters and print the resulting command line (so the dev can check if a given random permitted input makes sense).

### CODE REVIEW

File: localExec.py [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+
+Note: validating the schema with a validator is recommended first. This script does not check the descriptor with respect to the schema.
+      Only in the -e case are output files checked for, at the end
+
+Formats:
+A .json input file should look like:
+{
+  "inputs" : [
+    { "input_id_1" : "val1" },
+    { "input_id_2" : "val2" },
```
Comment (glatard) [inline localExec.py]: Are lists specified as JSON arrays?

Comment (glatard) [inline localExec.py]: Do we have to have only 1 value per input id? I think we should, but then we should write it.

File: localExec.py [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+A .csv input file should look like:
+input_id_1,val1
+input_id_2,val2
+...
+
+A string input should look like:
+python localExec.py desc.json -s 'in_id_1,val_1;in_id_2,val_2;...'
+
+Notes: pass lists by space-separated values [cannot be passed by --string]
+       pass flags with 'true' or 'false' (false is the same as not including it)
```
Comment (glatard) [inline localExec.py]: Examples of these 2 cases would be useful.

File: localExec.py [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+      sys.stderr.write( 'OS Error during attempted execution!' )
+      raise e
+    except ValueError as e:
+      sys.stderr.write( 'Input Value Error during attempted execution!' )
+      raise e
+    else:
+      exit_code = process.wait()
+      # Report exit status
+      print('---/* End program output */---\nCompleted execution (exit code: ' + str(exit_code) + ')')
+      time.sleep(0.5) # Give the OS a (half) second to finish writing
```
Comment (glatard) [inline localExec.py]: Is this really required after the wait at line 80?

File: localExec.py [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+          # and not search for them if the key is not present
+          val = '' if not (kid in self.in_dict.keys()) else self.in_dict[kid]
+          template = template.replace(key,val)
+        # Look for the target file
+        exists = os.path.exists( template )
+        # Note whether it could be found or not
+        isOptional = outfile['optional'] if 'optional' in outfile.keys() else False
+        s1 = 'Optional' if isOptional else 'Required'
+        s2 = '' if exists else 'not '
+        print("\t"+s1+" output file \'"+outfile['name']+"\' was "+s2+"found at "+template)
```
Comment (glatard) [inline localExec.py]: Could print this as an error if Required output file was not found.

File: localExec.py [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+
+  # Generates a random input parameter set that follows the given constraints
+  def _randomFillInDict(self):
+    # Helpers for generating random numbers, strings, etc...
+    nd, nl   = 2, 5 # Number of random chars to add to things, or max list items to have
+    randDigs = lambda: ''.join(rnd.choice(string.digits) for _ in range(nd))
+    randFile = lambda: 'f_' + randDigs() + rnd.choice(['.csv','.tex','.j','.cpp','.m','.mnc','.nii.gz'])
+    randStr  = lambda: 'str_' + ''.join(rnd.choice(string.digits+string.ascii_letters) for _ in range(nd))
+    def randNum(p): # generate random number that satisfies the param constraints
```
Comment (glatard) [inline localExec.py]: Could better comment on what p is here.

File: localExec.py [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+      if self.safeGet( i , 'exclusive-minimum' ): minv += (1 if isInt else 0.0001)
+      if self.safeGet( i , 'exclusive-maximum' ): maxv -= (1 if isInt else 0.0001)
+      return (rnd.randint(minv,maxv) if isInt else round(rnd.uniform(minv,maxv),nd) )
+    def paramSingle(prm): # generate a random parameter value based on the input type
+      if prm['type']=='Enum':   return rnd.choice( self.safeGet(prm['id'],'enum-value-choices') )
+      if prm['type']=='File':   return randFile()
+      if prm['type']=='String': return randStr()
+      if prm['type']=='Number': return randNum(prm)
+      if prm['type']=='Flag':   return rnd.choice(['true','false'])
+    def makeParam(prm): # generate random single or list params based on inputs
```
Comment (glatard) [inline localExec.py]: why would you generate a list if prm is not a list?

**Comment by glatard** [comment]
This is brilliant! A few minor comments in addition to the ones above:
* The executor could be added to README.md.
* When a Docker image is specified, option "-e" should run the command line in the docker container.

File: localExec.py [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+python localExec.py desc.json -s 'in_id_1,val_1;in_id_2,val_2;...'
+
+Notes: pass lists by space-separated values [cannot be passed by --string]
+       pass flags with 'true' or 'false' (false is the same as not including it)
+
+'''
+
+import argparse, os, sys, json, random as rnd, string, math, subprocess, time
+
+# Executor class
```
Comment (glatard) [inline localExec.py]: A few more lines of comment would be useful to explain what this class represents and does.

File: localExec.py [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+    self.reqsOf     = lambda t: self.safeGet(t,"requires-inputs") or []
+
+  # Attempt execution
+  def execute(self):
+    command = self.cmdLine[0]
+    print('Attempting execution of command:\n\t' + command + '\n---/* Start program output */---')
+    # Run command
+    try:
+      # Note: invokes the command through the shell (potential injection dangers)
+      process = subprocess.Popen( command , shell=True )
```
Comment (glatard) [inline localExec.py]: The command should be executed in a docker container when one is available.

File: localExec.py [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+      exit_code = process.wait()
+      # Report exit status
+      print('---/* End program output */---\nCompleted execution (exit code: ' + str(exit_code) + ')')
+      time.sleep(0.5) # Give the OS a (half) second to finish writing
+      # Check for output files (note: the path-template can contain command-line-keys
+      print('Looking for output files:')
+      for outfile in self.desc_dict['output-files']:
+        # Generate correct path template based on input arguments
+        template = outfile['path-template']
+        for (kid,key) in [(q['id'],q['command-line-key']) for q in self.inputs]:
```
Comment (glatard) [inline localExec.py]: what about renaming q for 'input'?

File: localExec.py [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+          # Blanks are entered if template key is not present
+          # Note: may wish to detect output files whose paths depend on particular inputs
+          # and not search for them if the key is not present
+          val = '' if not (kid in self.in_dict.keys()) else self.in_dict[kid]
+          template = template.replace(key,val)
```
Comment (glatard) [inline localExec.py]: Shouldn't this template resolution be done once and for all when the command line is built?

File: localExec.py [hunk 11]
```diff
@@ -0,0 +1,424 @@
...
+        # Look for the target file
+        exists = os.path.exists( template )
+        # Note whether it could be found or not
+        isOptional = outfile['optional'] if 'optional' in outfile.keys() else False
+        s1 = 'Optional' if isOptional else 'Required'
+        s2 = '' if exists else 'not '
+        print("\t"+s1+" output file \'"+outfile['name']+"\' was "+s2+"found at "+template)
+
+  # Generates a random input parameter set that follows the given constraints
```
Comment (glatard) [inline localExec.py]: Explain where the constraints are specified?

File: localExec.py [hunk 12]
```diff
@@ -0,0 +1,424 @@
...
+      if minv is None and maxv is None: minv, maxv = defaultMin, defaultMax
+      elif minv is None and not (maxv is None): minv = maxv + defaultMin
+      elif not (minv is None) and maxv is None: maxv = minv + defaultMax
+      # Coerce the min/max to the proper number type
+      if isInt: minv, maxv = roundTowardsZero(minv), roundTowardsZero(maxv)
+      else: minv, maxv = float(minv), float(maxv)
+      # Apply exclusive boundaries if any
+      if self.safeGet( i , 'exclusive-minimum' ): minv += (1 if isInt else 0.0001)
+      if self.safeGet( i , 'exclusive-maximum' ): maxv -= (1 if isInt else 0.0001)
+      return (rnd.randint(minv,maxv) if isInt else round(rnd.uniform(minv,maxv),nd) )
```
Comment (glatard) [inline localExec.py]: Unclear what nd is here, the comment at line 106 doesn't help much.

File: localExec.py [hunk 13]
```diff
@@ -0,0 +1,424 @@
...
+
+  # Generates a random input parameter set that follows the given constraints
+  def _randomFillInDict(self):
+    # Helpers for generating random numbers, strings, etc...
+    nd, nl   = 2, 5 # Number of random chars to add to things, or max list items to have
+    randDigs = lambda: ''.join(rnd.choice(string.digits) for _ in range(nd))
+    randFile = lambda: 'f_' + randDigs() + rnd.choice(['.csv','.tex','.j','.cpp','.m','.mnc','.nii.gz'])
+    randStr  = lambda: 'str_' + ''.join(rnd.choice(string.digits+string.ascii_letters) for _ in range(nd))
+    def randNum(p): # generate random number that satisfies the param constraints
+      i, defaultMin, defaultMax = p['id'], -50, 50
```
Comment (glatard) [inline localExec.py]: rename 'i' to 'param_id'?

File: localExec.py [hunk 14]
```diff
@@ -0,0 +1,424 @@
...
+      isInt = self.safeGet( i , 'integer' )
+      roundTowardsZero = lambda x: int(math.copysign(1,x)*int(abs(x)))
+      # Assign random values to min and max, unless they have been specified
+      minv, maxv = self.safeGet( i , 'minimum' ), self.safeGet( i , 'maximum' )
+      if minv is None and maxv is None: minv, maxv = defaultMin, defaultMax
+      elif minv is None and not (maxv is None): minv = maxv + defaultMin
+      elif not (minv is None) and maxv is None: maxv = minv + defaultMax
+      # Coerce the min/max to the proper number type
+      if isInt: minv, maxv = roundTowardsZero(minv), roundTowardsZero(maxv)
```
Comment (glatard) [inline localExec.py]: Isn't there any built-in function in python to do what 'roundTowardsZero' does?

File: localExec.py [hunk 15]
```diff
@@ -0,0 +1,424 @@
...
+    def makeParam(prm): # generate random single or list params based on inputs
+      mn = self.safeGet(prm['id'], 'min-list-entries') or 2
+      mx = self.safeGet(prm['id'], 'max-list-entries') or nl
+      isList = self.safeGet(prm['id'], 'list') or False
+      return ' '.join(str(paramSingle(prm)) for _ in range(rnd.randint(mn,mx))) if isList else paramSingle(prm)
+    def disablersOf(tid): # Returns a list of the ids of parameters that disable the input parameter
+      return map(lambda p: p[0], # p[0] is the id
+             filter(lambda t: tid['id'] in t[1], # t[1] is the disables list
+             map(lambda p: (p['id'], self.safeGet(p['id'],'disables-inputs') or []), self.inputs) ))
+    def mutReqs( targ ): # Returns the list of mutually requiring parameters of the target
```
Comment (glatard) [inline localExec.py]: could rename 'targ' to a more explicit name (and other variables too).

File: localExec.py (hunk 1 above)
Comment (ttaa9) [inline localExec.py]: No they are space-separated right now, but I think a json array would be a better idea actually

File: localExec.py (hunk 6 above)
Comment (ttaa9) [inline localExec.py]: makeParam generates a single param or a list param depending on whether the schema says it is a list or not. If prm is not a list, it does not generate list; it only calls paramSingle once.

File: localExec.py (hunk 10 above)
Comment (ttaa9) [inline localExec.py]: I'm not exactly sure what you mean. Here, the template is the "path-template" of the output file. (Separate from the command-line template). I don't bother doing the template resolution of the output file if I am just printing the command-line.

File: localExec.py (hunk 14 above)
Comment (ttaa9) [inline localExec.py]: Sadly, not one that I could find :( Even C has one apparently...

File: localExec.py (hunk 3 above)
Comment (ttaa9) [inline localExec.py]: Probably not most of the time, but I remember needing to sleep for almost 10 seconds in some crazier cases (not related to Boutiques). The `wait` only waits for the process to complete, not for the OS to complete all the related queued write tasks, I guess.

File: localExec.py (hunk 6 above)
Comment (glatard) [inline localExec.py]: ok!

File: localExec.py (hunk 10 above)
Comment (glatard) [inline localExec.py]: OK.

File: localExec.py (hunk 3 above)
Comment (glatard) [inline localExec.py]: OK!

**Pull request merged**